Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp

```diff
@@ -1102,29 +1303,44 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::keys)
     return ArrayIterator::create(global_object, this_object, Object::PropertyKind::Key);
 }
 
-static void recursive_array_flat(VM& vm, GlobalObject& global_object, Array& new_array, Object& array, double depth)
+// 23.1.3.10.1 FlattenIntoArray ( target, source, sourceLen, start, depth [ , mapperFunction [ , thisArg ] ] ), https://tc39.es/ecma262/#sec-flattenintoarray
+static size_t flatten_into_array(VM& vm, GlobalObject& global_object, Array& new_array, Object& array, size_t target_index, double depth, Function* mapper_func = {}, Value this_arg = {})
 {
+    VERIFY(!mapper_func || (!this_arg.is_empty() && depth == 1));
     auto array_length = length_of_array_like(global_object, array);
     if (vm.exception())
-        return;
+        return -1;
 
     for (size_t j = 0; j < array_length; ++j) {
+        auto value_exists = array.has_property(j);
+        if (vm.exception())
+            return -1;
+        if (!value_exists)
+            continue;
         auto value = array.get(j);
         if (vm.exception())
-            return;
+            return -1;
+
+        if (mapper_func) {
+            value = vm.call(*mapper_func, this_arg, value, Value(j), array.value_of());
```

Comment:
```suggestion
            value = vm.call(*mapper_func, this_arg, value, Value(j), array);
            if (vm.exception())
                return {};
```

Reply:
Array is a reference here so this does not work. Should i use `&array` or `array.value_of()`?

Reply:
Ah right, &array then

Reply:
Changed. I think i finally fixed the entire function then